File: services/cart/src/lib/metrics.ts
import client from 'prom-client';

export const registry = new client.Registry();
client.collectDefaultMetrics({ register: registry });

export const httpRequestCount = new client.Counter({
  name: 'http_requests_total',
  help: 'Total number of HTTP requests received',
  labelNames: ['method', 'route', 'status_code'] as const,
  registers: [registry],
});

export const httpRequestDuration = new client.Histogram({
  name: 'http_request_duration_seconds',
  help: 'HTTP request duration in seconds',
  labelNames: ['method', 'route', 'status_code'] as const,
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5],
  registers: [registry],
});

export const httpErrorCount = new client.Counter({
  name: 'http_errors_total',
  help: 'Total number of HTTP requests that resulted in an error (status >= 500)',
  labelNames: ['method', 'route', 'status_code'] as const,
  registers: [registry],
});

File: services/order/internal/metrics/metrics.go
// Package metrics defines the Prometheus collectors exposed at GET /metrics.
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	// RequestsTotal counts every HTTP request handled by the service,
	// labeled by method, route pattern, and status code.
	RequestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "order_http_requests_total",
			Help: "Total number of HTTP requests processed by the order service.",
		},
		[]string{"method", "route", "status"},
	)

	// RequestDuration observes request handling latency in seconds,
	// labeled by method and route pattern.
	RequestDuration = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "order_http_request_duration_seconds",
			Help:    "Histogram of HTTP request durations in seconds.",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"method", "route"},
	)

	// ErrorsTotal counts requests that resulted in a 5xx (or handler-level
	// error) response, labeled by method and route pattern.
	ErrorsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "order_http_errors_total",
			Help: "Total number of HTTP requests that resulted in an error response.",
		},
		[]string{"method", "route", "status"},
	)

	// EventsProducedTotal counts Kafka events published, labeled by topic.
	EventsProducedTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "order_kafka_events_produced_total",
			Help: "Total number of Kafka events published by the order service.",
		},
		[]string{"topic", "eventType"},
	)

	// EventsConsumedTotal counts Kafka events consumed, labeled by topic.
	EventsConsumedTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "order_kafka_events_consumed_total",
			Help: "Total number of Kafka events consumed by the order service.",
		},
		[]string{"topic", "eventType"},
	)
)
